src/sources/prometheus/scrape.rs

```diff
@@ -529,6 +567,75 @@ mod test {
         }
     }
 
+    #[tokio::test]
+    async fn test_prometheus_request_query() {
+        let in_addr = next_addr();
+
+        let dummy_endpoint = warp::path!("metrics").and(warp::query::raw()).map(|query| {
+            format!(
+                r#"
+                    promhttp_metric_handler_requests_total{{query="{}"}} 100 1612411516789
+                "#,
+                query
+            )
+        });
+
+        tokio::spawn(warp::serve(dummy_endpoint).run(in_addr));
+
+        let query = HashMap::from([
+            ("key1".to_string(), vec!["val2".to_string()]),
+            (
+                "key2".to_string(),
+                vec!["val1".to_string(), "val2".to_string()],
+            ),
+        ]);
+
+        let config = PrometheusScrapeConfig {
+            endpoints: vec![format!("http://{}/metrics?key1=val1", in_addr)],
+            scrape_interval_secs: 1,
+            instance_tag: Some("instance".to_string()),
+            endpoint_tag: Some("endpoint".to_string()),
+            honor_labels: false,
+            query: Some(query.clone()),
+            auth: None,
+            tls: None,
+        };
+
+        let (tx, rx) = SourceSender::new_test();
+        let source = config
+            .build(SourceContext::new_test(tx, None))
+            .await
+            .unwrap();
+
+        tokio::spawn(source);
+        tokio::time::sleep(Duration::from_secs(1)).await;
+
+        let events = test_util::collect_ready(rx).await;
+        assert!(!events.is_empty());
+
+        let metrics: Vec<_> = events
+            .into_iter()
+            .map(|event| event.into_metric())
+            .collect();
+
+        for metric in metrics {
+            if let Some(q) = metric.tag_value("query") {
+                // note: request query has keys in arbitrary order
+                let mut parsed_query: HashMap<String, Vec<String>> = HashMap::new();
+                for (k, v) in url::form_urlencoded::parse(q.as_bytes()) {
+                    parsed_query
+                        .entry(k.to_string())
+                        .or_insert(Vec::new())
+                        .push(v.to_string());
+                }
+                for (k, l) in &query {
+                    assert!(parsed_query.contains_key(k));
+                    assert_eq!(parsed_query[k].clone().sort(), l.clone().sort());
+                }
```

Comment:
I also just realised that I should check the map both ways, to account for extra keys being returned in the parsed query. The validation should be 1:1 of both maps. I hope there is a better way to do this assertion.

I also just saw that the CI didn't like this test. Will check it out.
But again, feedback is highly welcome in this area :)